include/mesh/mesh_base.h

```diff
@@ -263,6 +263,29 @@ class MeshBase : public ParallelObject
    */
   void set_elem_dimensions(const std::set<unsigned char> & elem_dims);
 
+  /**
+   * Tabulate a user-defined "code" for elements which belong to the element sets
+   * specified in \p id_set. Also sets up the inverse mapping, so that if one knows
+   * all the element sets an Elem belongs to, one can look up the corresponding code.
+   */
+  void add_elemset_code(dof_id_type code, const std::set<elemset_id_type> & id_set);
+
+  /**
+   * Determines the number of unique elemset ids which have been added
+   * via add_elemset_code() calls by looping over the
+   * _elemset_codes_inverse_map and counting them.
+   */
+  unsigned int n_elemsets() const;
+
+  /**
+   * Look up the element sets for a given elemset code and
+   * vice-versa. The elemset must have been previously stored by
+   * calling add_elemset_code(). If no such code/set is found, returns
+   * the empty set or DofObject::invalid_id, respectively.
+   */
+  void get_elemsets(dof_id_type elemset_code, std::set<elemset_id_type> & id_set_to_fill) const;
```

Comment:
Should we just return rather than fill in new APIs, now that C++17 mandates copy/move construction?

Reply:
I thought about this, but I generally expect this function to be called for each Elem in an Elem-loop... so I think that would still construct/destroy one std::set per loop iteration (?) whereas with this you could construct the std::set outside the loop and reuse it. Or maybe you are saying that

```
std::set<elemset_id_type> s;
for (elem loop...)
{
  s = get_elemsets(elem->get_extra_integer(elemset_index));
}
```
would move-assign `s` each time, giving the same effect?

Reply:
Oh, that's an interesting point.

I think the construction *doesn't* get elided here, because that `s =` is an assignment rather than an initialization.  It will be a move assignment, but that move assignment won't have the same effect in any `std::set` implementation that reuses node memory, but I'm not sure if there *is* any `std::set` implementation that reuses memory.  If there is then pulling the `set` construction out of the loop would be helpful, but if there isn't (and I think there isn't) then it's not.  But then if we were to switch to an underlying sorted `vector` or an `unordered_set` we're back to "helpful" again.

Reply:
I ended up leaving this one as pass-by-reference for now since I think it's the familiar approach, and its effectiveness does not depend on whatever type we might eventually change `MeshBase::elemset_type` to.